LoopKit/InsulinKit/InsulinMath.swift

```diff
@@ -84,21 +84,35 @@ extension DoseEntry {
         }
     }
 
-    func trim(from start: Date? = nil, to end: Date? = nil) -> DoseEntry {
-        guard unit == .unitsPerHour else {
-            return self
-        }
+    func trimmed(from start: Date? = nil, to end: Date? = nil, syncIdentifier: String? = nil) -> DoseEntry {
+
+        let originalDuration = endDate.timeIntervalSince(startDate)
 
         let startDate = max(start ?? .distantPast, self.startDate)
+        let endDate = max(startDate, min(end ?? .distantFuture, self.endDate))
+
+        var trimmedDeliveredUnits: Double? = deliveredUnits
+        var trimmedValue: Double = value
+
+        if originalDuration > .ulpOfOne && (startDate > self.startDate || endDate < self.endDate) {
+            let updatedActualDelivery = unitsInDeliverableIncrements * (endDate.timeIntervalSince(startDate) / originalDuration)
```

Comment:
Yep, this was a conscious decision to represent things this way. It will take a bit of explaining. In the "annotation" phase, when we take reconciled doses and map them to a basal schedule, we need to split them up along basal schedule boundaries, so we can compute insulin activity relative to the basal schedule.

I believe your question is basically asking if we can apply a pulsed based view to this dose and split it with the pulses in the right place. In theory, we could, but the how gets complicated.  This function is a ways downstream from dose reconciliation and the PumpManager, which has the knowledge of how to do that. Could we give the PumpManager this responsibility?  We could, but the idea of considering a dose as relative to a basal schedule is really a Loop algorithm consideration; we could very well do away with the idea of relative (negative insulin!?!) iob altogether, and have discussed doing so many times.  So I think this work is best kept outside of pump managers.

How downstream is this? It's PumpManager -> Loop -> DoseStore.pumpEvents -> HealthKit -> InsulinDeliveryCache. Asking the current pumpmanager about a dose that was retrieved from HK doesn't work. Recording a pulse schedule with each dose would work, but that's complex to flow that through the above chain, and a pretty good example of overprecision, considering the actual current physics of pump based insulin delivery.

Lastly, what are the possible errors in looking in doing the math this way? We're talking a fraction of a pulse, and most importantly, an error that does not accumulate over time.

Reply:
:+1: Pretty much what I expected the case to be.